doc/neps/nep-0049.rst

```diff
@@ -143,6 +143,9 @@ NumPy C-API functions
 
     The consumer of the `PyDataMemAllocator` interface must keep track of ``size`` and make sure it is
     consistent with the parameter passed to the ``(m|c|re)alloc``  functions.
+    NumPy itself will violate this requirement when the shape of the requested
+    array contains a ``0``, so authors of PyDataMemAllocators should relate to
+    the ``size`` parameter as a best-guess.
```

Comment:
Shouldn't we just fix our call to free to be consistent here?

Reply:
Or is computing the actual allocated size something that is too expensive to do in every array deallocation?

Reply:
It is not so simple. For instance `np.fromfile` calls `array_fom_text`, which calls `PyArray_NewFromDescr()` to allocate a buffer. Then it reads from the file, assigning to the result, and reallocating if it runs out of room. How will free know what size to release? 

<details><summary>code from array_from_text</summary>

https://github.com/numpy/numpy/blob/df6d2600c51502e1877aac563658d0616a75c5e5/numpy/core/src/multiarray/ctors.c#L3370-L3420

Reply:
I think we can solve these edge cases, but I would prefer to do it once the NEP and first implementation PR are merged.

Reply:
It looks to me like that function always trims the array before there is a chance of it being freed:

https://github.com/numpy/numpy/blob/df6d2600c51502e1877aac563658d0616a75c5e5/numpy/core/src/multiarray/ctors.c#L3426

I suppose if that reallocation fails though, then the free call has no chance of knowing the size of the array; but maybe that's a bug. If that code updated `PyArray_DIMS` at the same time as it updated `data`, then there would be no such problem; the deallocator for `ndarray` could just take the product of its non-zero dimensions to determine the size (which I think is the logic used for allocation?

Reply:
As far as I can tell the logic for the first allocation uses `size` elements, which is initialized to `FROM_BUFFER_SIZE` and is not connected to the product of the dimensions.

Reply:
Yes, your reading is correct - but at that point in time, the shape of the array is `(size,)` so the size is still linked to the the product of the dimensions `(size,)`.

Reply:
OK, it was a bug in the `shift_realloc` function in the test. When reallocating it did not change the hidden header to reflect the new size. Now I am down to what do with
```
a = np.array([])
a.shape = (3, 2, 1, 0)
```

If it were `a = np.empty([3, 2, 1, 0])`, it would reach `PyArray_NewFromDescr_int` and ignore the 0, allocating 48 bytes. When it comes to deallocating, the routine does not know that the original array was `np.array([])`

Reply:
In c331dc4a7ac I think I fixed the mismatches.

Reply:
You're right that this is a troublesome example:
```python
a = np.array([])
a.shape = (3, 2, 1, 0)
```
The options I can think of are:
* Revive that PR that makes all 0-element arrays have an allocation of 1 element / 1 byte
* Do a realloc even though it's stupid
* Add the caveat you had before
* Make such code an error

The fixes in c331dc4a7ac68fcfe67db75c1b3f6a3c531dc12f look like they fix most of the corner cases, but if there are any left we should keep the caveat around.

Reply:
CI (except for circleCI doc building, which seems to have an outdated image) is passing. Can we move forward with the NEP so that we can get it approved and merge the PR?

Reply:
Ahh, OK, I see in #15780 and #15788 that the issue is not so simple. I will revert the changes in the implementation PR for arrays with 0 in their shape in c331dc4.